core/src/main/java/org/apache/brooklyn/core/OsgiLauncher.java

```diff
@@ -0,0 +1,557 @@
+import org.apache.brooklyn.core.mgmt.internal.LocalManagementContext;
+import org.apache.brooklyn.core.mgmt.internal.ManagementContextInternal;
+import org.apache.brooklyn.core.mgmt.persist.BrooklynMementoPersisterToObjectStore;
+import org.apache.brooklyn.core.mgmt.persist.BrooklynPersistenceUtils;
+import org.apache.brooklyn.core.mgmt.persist.PersistMode;
+import org.apache.brooklyn.core.mgmt.persist.PersistenceObjectStore;
+import org.apache.brooklyn.core.mgmt.rebind.PersistenceExceptionHandlerImpl;
+import org.apache.brooklyn.core.mgmt.rebind.RebindManagerImpl;
+import org.apache.brooklyn.core.objs.BrooklynTypes;
+import org.apache.brooklyn.core.server.BrooklynServerConfig;
+import org.apache.brooklyn.core.server.BrooklynServerPaths;
+import org.apache.brooklyn.util.core.ResourceUtils;
+import org.apache.brooklyn.util.exceptions.Exceptions;
+import org.apache.brooklyn.util.exceptions.FatalConfigurationRuntimeException;
+import org.apache.brooklyn.util.exceptions.FatalRuntimeException;
+import org.apache.brooklyn.util.exceptions.RuntimeInterruptedException;
+import org.apache.brooklyn.util.guava.Maybe;
+import org.apache.brooklyn.util.io.FileUtil;
+import org.apache.brooklyn.util.os.Os;
+import org.apache.brooklyn.util.stream.Streams;
+import org.apache.brooklyn.util.text.Strings;
+import org.apache.brooklyn.util.time.Duration;
+import org.apache.brooklyn.util.time.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Initializer for brooklyn-core when running in an OSGi environment.
+ *
+ * Temporarily here; should be totally contained in blueprint beans' init-methods.
+ */
+public class OsgiLauncher {
+
+    private static final Logger log = LoggerFactory.getLogger(OsgiLauncher.class);
+
+    private ManagementContext managementContext;
+
+    private boolean ignoreCatalogErrors = true;
+    private boolean ignorePersistenceErrors = true;
+
+    private HighAvailabilityMode highAvailabilityMode = HighAvailabilityMode.DISABLED;
+
+    private PersistMode persistMode = PersistMode.DISABLED;
+    private String persistenceDir = null;
+    private String persistenceLocation = null;
+    private Duration persistPeriod = Duration.ONE_SECOND;
+
+    public void start() {
+        // Configure launcher
+//        ShutdownHandler shutdownHandler = new AppShutdownHandler();
+        try {
+//            initBrooklynProperties();
+            initCatalog();
+
+            // TODO
+//            launcher.stopWhichAppsOnShutdown(stopWhichAppsOnShutdownMode);
+//            launcher.shutdownHandler(shutdownHandler);
+//
+//            computeAndSetApp(launcher, utils, loader);
+//            customize(launcher);
+//        } catch (FatalConfigurationRuntimeException e) {
+//            throw e;
+//        } catch (Exception e) {
+//            throw new FatalConfigurationRuntimeException("Fatal error configuring Brooklyn launch: " + e.getMessage(), e);
+//        }
+//
+//        try {
```

Comment:
I am unsure in which direction brooklyn will evolve within karaf, so I left all these commented out for review. I'll refer to each aspect of the code commented above here:

- `initBrooklynProperties()` - right now it is not on the table, since we're trying to preserve the existing functionality within the normal launcher, but `brooklyn.properties` should disappear in favor of karaf's `ConfigAdmin` interface. By default this looks up `cfg` files in karaf's `etc` folder, but these are actually normal property files. Also, karaf can automatically inject properties from these `cfg` files in various java beans and notify when the files change on disk. Within karaf cellar we get these changes distributed on all nodes within a cluster group. Furthermore, going this path allows multiple instances of karaf on the same machine, easing development and management.
- *stopping apps on shutdown* does not really makes sense in karaf, where `brooklyn-core` can be stopped or started with various purposes (upgrades, cellar nodes being added or removed from the cluster etc). We might envision a concept of stopping all existing apps independently of brooklyn's lifecycle. Would that be acceptable?
- the same applies to `computeAndSetApp` from the command-line. One can envision karaf shell commands that take care of this. Or we could try and preserve the existing command-line functions, but these will be awkward to match the bundle lifecycle in the container (when do we really say it "started"? what action do we interpret as "shutdown"?)
- `customize(launcher)` - I think this applied command-line arguments to the launcher, so it ties to the above discussion. On the other hand, any change to karaf's `etc` folder will reload the configuration, so any customization is envisionable through this mechanism. Also, karaf shell commands can be used to change behavior as well.

I hope you will review the ideas above and help decide what to do in the short term and the long term.